### ISSUE TITLE
[FEATURE] Algebraic transformations in the preprocessor

### ISSUE THREAD

**Issue by konnov** [informalsystems/apalache#480]
There are a number of expressions that can be significantly simplified by using logical and algebraic equivalences.

**Example 1:**

```tla
x \in X \union Y
```

can be simplified to

```tla
x \in X /\ x \in Y
```

**Example 2:**

```tla
\E x \in a..b: P
```

can be simplified as

```tla
\E x \in Int: a <= x /\ x <= b /\ P
```

We are doing some of simplifications in the preprocessors already, but we should do that more systematically by introducing a rewriting framework. By doing so, we should eliminate a large class of issues, where our encoding is suboptimal.

**Comment by konnov** [comment]
By a rewriting framework I meant something like [the worklist algorithm](https://en.wikipedia.org/wiki/Data-flow_analysis#The_work_list_approach). The point is to avoid unnecessary recursive calls when subexpressions were not optimized. However, I am not sure about how to do that in Scala.

**Comment by gabrielamafra** [comment]
I'm not capturing the necessity of this type of algorithm here. My understanding is that we never need to re-simplify/optimize subexpressions of an expression we are rewriting, as long as we always rewrite using a bottom-up approach and never change the contents of a subexpression while rewriting. This is how `ConstSimplifier` is working: there is a single deep recursion and anything that changes halfway of the recursion is simplified before continuing to recurse and simplify its parent.

We could take a look at other rewrite implementations besides `ConstSimplifier` and check that they can work under this 2 assumptions and, if so, make sure that new nodes (created from a rewrite rule) are shallow simplified/optimized. This could be a first step towards unifying all rewrite implementations as suggested in https://github.com/informalsystems/apalache/issues/138

**Comment by konnov** [comment]
Right. We don't need any special framework to implement this kind of preprocessing. It would be great, if we could call different optimizers when one of the optimizers has changed a subexpression. A naive fixpoint approach applied to a whole expression over and over again would be really slow. I think we would need some form of recursion that applies different optimizations to subexpressions and once nothing new has been produced, it produces the result upwards to the parent.
